FEA CSR support for all `DistanceMetric`

jjerphan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Precedes support for fused sparse-dense datasets for `PairwiseDistancesReductions` (see https://github.com/scikit-learn/scikit-learn/issues/22587)


#### What does this implement/fix? Explain your changes.

This implements supports distance metric computation for CSR data.

Importantly, this:
 - define `DistanceMetric.{dist_csr,rdist_csr}` (adapted versions of `DistanceMetric.{dist,rdist}`) for CSR data (see the `pxd` file for the definition).
 - implement `DistanceMetric.{dist_csr,rdist_csr}` for all the `DistanceMetric` excepted `PyFuncDistance`.
 - this uses a indices wrapping to be able to support the sparse-dense, dense-sparse cases and to be robust to explicit zeros representation with a minimal memory footprint

#### Any other comments?

Additional changes:
 - use np.float64 for extra datastructures in all the cases for best precisions (namely, precision matrices, weights vectors, work vectors)
 - minor reformatting in implementation and minor renaming in tests for consistency
 - additional private python method to allow testing those interfaces
 - mahalanobis now tested using an adapted tolerance for its tests cases

ℹ️
 The +2000 diff really is due to logic duplication. It would be +600 if it were entirely factorised. Yet, I do not see how we can remove this easily without loosing performance and without additional costly indirection. Hence this choice comes for performance at the cost of maintainability.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->


### TODO:

- [x] implement sparse support for Haversine + dedicated tests
- [x] check that the float32 -> float64 upcasts are done at the right location to preserve numerical stability

jjerphan committed:
MAINT Implement CSR support for all DistanceMetric

jjerphan commented:
Not sure of why is fails for `Linux_Docker debian_atlas_32bit`. :thinking:

jjerphan committed:
Merge branch 'main' into maint/dist-metrics-csr-support

jjerphan committed:
TST Remove useless guard

jjerphan committed:
TST Skip JaccardDistance on 32bit architecture

jjerphan commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -104,36 +105,52 @@ def test_cdist(metric_param_grid, X, Y):
-    D_true = cdist(X_bool, Y_bool, metric)
+    D_scipy_cdist = cdist(X_bool, Y_bool, metric)
+
+    if metric == "jaccard" and _IS_32BIT:
+        pytest.skip("Jaccard Distance on 32bit architecture is unstable.")
```
This is the skip for the 32bit architecture.

ogrisel replied:
I think would rather `xfail` instead of skip.

But I would not block the merge of this PR for such a problem which is unlikely to impact many users in practice.

jjerphan commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -104,36 +105,52 @@ def test_cdist(metric_param_grid, X, Y):
-    D_true = cdist(X_bool, Y_bool, metric)
+    D_scipy_cdist = cdist(X_bool, Y_bool, metric)
+
+    if metric == "jaccard" and _IS_32BIT:
+        pytest.skip("Jaccard Distance on 32bit architecture is unstable.")
```
This is the skip for the 32bit architecture.

jjerphan committed:
MAINT Define dtype alias for sparse matrices indices

jjerphan committed:
MAINT Do not shadow dtype names in Tempita templating

jjerphan committed:
fixup! MAINT Define dtype alias for sparse matrices indices

jjerphan committed:
TST Use cdist and pdist appropriately

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -445,23 +643,10 @@ cdef class DistanceMetric{{name_suffix}}:
+        if not issparse(X) and (Y is None or not issparse(Y)):
+            return self._pairwise_dense(X, Y)
+        if issparse(X) and (Y is None or issparse(Y)):
+            return self._pairwise_sparse(X, Y)
```
Actually it would be nicer to make this support the sparse x dense case by doing the trick to cast the dense array as a CSR using the tiled `np.arange()` indices.

jjerphan committed:
Fixups

jjerphan committed:
DOC Improve comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
MAINT Wrap of indptr values to support sparse-dense

This is kind of an hack for now.

IMO, it would be better to use a flatiter on a view if possible.
See discussions on: https://groups.google.com/g/cython-users/c/MR4xWCvUKHU

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -424,6 +570,86 @@ cdef class DistanceMetric{{name_suffix}}:
+        X_data = np.asarray(X_csr.data, dtype={{INPUT_DTYPE}})
+        X_indices = np.asarray(X_csr.indices, dtype=SPARSE_INDEX_TYPE)
+        X_indptr = np.asarray(X_csr.indptr, dtype=SPARSE_INDEX_TYPE)
+
```
```suggestion

        # To avoid introducing redundant implementations for the CSR x dense array
        # case, we wrap the dense array into a fake CSR datastructure and leverage
        # the existing code for the CSR x CSR case.
        # The true CSR representation of a dense array would require allocating
        # a Y_indices matrix of shape (n_samples, n_features) with repeated
        # contiguous integers from 0 to n_features - 1 on each row which would
        # be very wasteful from a memory point of view. Instead we only allocate
        # a single row and adapt the CSR x CSR routines to use a modulo operation
        # when accessing Y_indices in order to achieve the same result with having
        # to materialize the indices repetition explicitly.
```

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -496,6 +724,87 @@ cdef class EuclideanDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
+            DTYPE_t unsquared = 0
+
+        while i1 < x1_end and i2 < x2_end:
+            # In the case of dense vectors, indices are repeated in the normal
+            # representation for all the rows, replicating information.
+            # Hence indices can be stored once in a buffer of `n_features` elements
+            # instead of being stored in a buffer of
+            # `n_features × n_samples` elements with repetition.
+            # To support both representations, we wrap `indptr` values.
```
Let's centralize the detailed explanation in the _pairwise_sparse_dense caller.

```suggestion
            # Use the modulo-trick to implement support for CSR x dense array
            # with the CSR x CSR routine. See _pairwise_sparse_dense for more
            # details.
```

ogrisel replied:
Similar repeated comments should be updated accordingly.

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -424,6 +570,86 @@ cdef class DistanceMetric{{name_suffix}}:
+        n_Y, _ = Y.shape
+        Y_data = Y.reshape(-1)
+        Y_indices = np.arange(size, dtype=SPARSE_INDEX_TYPE)
+        Y_indptr = np.arange(stop=size*(n_Y + 1), step=size, dtype=SPARSE_INDEX_TYPE)
```
I believe this will fix the tests:

```suggestion
        Y_indptr = np.arange(
            start=0, stop=size*(n_Y + 1), step=size, dtype=SPARSE_INDEX_TYPE
        )
```

jjerphan committed:
Apply review comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
FIX Various corrections

ogrisel committed:
More interesting boolean data for tests

jjerphan committed:
FIX Make Jaccard, Hamming and Hashing robust to explicit zeros

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
FIX Make the other boolean DistanceMetric also robust to explicit zeros

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
TST Remove xfail for Jaccard on 32bit arch.

ogrisel commented:
We still need to review that all the upcasts are done at the safest location.

jjerphan committed:
Cast to np.float64_t where appropriate

lorentzenchr commented on sklearn/metrics/_dist_metrics.pxd.tp:
```diff
@@ -78,27 +78,82 @@ cdef class DistanceMetric{{name_suffix}}:
-    cdef DTYPE_t _rdist_to_dist(self, {{DTYPE_t}} rdist) nogil except -1
-
-    cdef DTYPE_t _dist_to_rdist(self, {{DTYPE_t}} dist) nogil except -1
+    cdef DTYPE_t dist(self,
+        const {{INPUT_DTYPE_t}}* x1,
+        const {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size,
+    ) nogil except -1
```
```suggestion
    cdef DTYPE_t dist(
        self,
        const {{INPUT_DTYPE_t}}* x1,
        const {{INPUT_DTYPE_t}}* x2,
        ITYPE_t size,
    ) nogil except -1
```
This would be more like black.

lorentzenchr commented on sklearn/metrics/_dist_metrics.pxd.tp:
```diff
@@ -78,27 +78,82 @@ cdef class DistanceMetric{{name_suffix}}:
+        ITYPE_t size,
+    ) nogil except -1
+
+    cdef DTYPE_t csr_dist(self,
```
Shall we name it
1. `csr_dist` or
2. `dist_csr`

and so on? I'm a little bit more fore `dist_csr`.

lorentzenchr commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -103,37 +104,73 @@ def test_cdist(metric_param_grid, X, Y):
 
         dm = DistanceMetricInterface.get_metric(metric, **kwargs)
+
+        # DistanceMetric.pairwise must be consistent
+        # on all combinations of format in {sparse, dense}².
```
```suggestion
        # DistanceMetric.pairwise must be consistent for all
        # combinations of formats in {sparse, dense}.
```
How about looping over those combinations?

jjerphan replied:
Having it unrolled allow identifying in case of failure which combination fails directly.

Do you think we can get this information if looping on combinations?

lorentzenchr replied:
🤔 Fixtures or let it be as is.

jjerphan replied:
At the moment, I do not see how one can use fixtures to perform it.
Feel free to do a PR on top of this one.

lorentzenchr replied:
So let's keep it as is.

lorentzenchr reviewed (commented):
Partial review. I left out the big part of `_dist_metrics.pyx.tp`.

jjerphan committed:
Rename methods and correctly format their signatures

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

jjerphan committed:
fixup! TST Remove xfail for Jaccard on 32bit arch.

jjerphan committed:
FEA CSR support for HaversineDistance

ogrisel reviewed (commented):
Thanks @jjerphan! The design, docstrings, tests and code look good to me.

The only remaining question is the need to update the assertions in the dense test (see below) because we now to the upcasting before the element-wise differences (both in the dense and sparse cases) while previously we would do the the upcasting to float64 right after the element wise differences:

jjerphan committed:
Fix typo

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -507,31 +845,39 @@ cdef class SEuclideanDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
-            tmp = <DTYPE_t> (x1[j] - x2[j])
-            d += <DTYPE_t> (tmp * tmp / self.vec[j])
+            tmp = (<DTYPE_t>(x1[j]) - <DTYPE_t>(x2[j]))
+            d += (tmp * tmp / self.vec[j])
```
This is the new up-casting strategy (upcasting before the difference, rather than after) that is likely the cause of the change required in `test_distance_metrics_dtype_consistency`.

ogrisel commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -166,25 +209,24 @@ def test_distance_metrics_dtype_consistency(metric_param_grid):
         D64 = dm64.pairwise(X64, Y64)
         D32 = dm32.pairwise(X32, Y32)
-        assert_allclose(D64, D32)
+        assert_allclose(D64, D32, rtol=1e-5)
```
Actually the default for the scikit-learn `assert_allclose` when one of the args is float32 is`rtol=1e-4` so this change in the test makes the test more strict. I misinterpreted this change.

As discussed in real life, doing an explicit upcast before the element-wise differences should not improve the precision and might make the code run slightly slower.

jjerphan replied:
Explicit cast has been removed and a comment for this custom `rtol` has been added in  563e3590c4bb5b3e3a6de4f42bc2e6f56a4769dc.

jjerphan committed:
Do not upcast to 64bit yet keep the same precision

jeremiedbb commented on sklearn/metrics/tests/test_dist_metrics.py:
```diff
@@ -166,25 +209,28 @@ def test_distance_metrics_dtype_consistency(metric_param_grid):
         D64 = dm64.pairwise(X64)
         D32 = dm32.pairwise(X32)
-        assert_allclose(D64, D32)
+
+        # The original rtol is 1e-4 in of presence of float32.
+        # We increase the tolerance to 1e-6 to be stricter.
+        rtol = 1e-6
+        assert_allclose(D64, D32, rtol=rtol)
```
But why not leaving `rtol` to its default value ?
It's kind of lucky that it passes for 1e-6 but it would probably fail if we tweak the datasets a bit.

ogrisel replied:
Actually this will fail both most datasets: both `D64.dtype` and `D32.dtype` are `np.float64` so the default `rtol` will be 1e-7 we need to set rtol explicitly to a lower value such as 1e-4 (to be lax) or 1e-5/1e-6 to be stricter after running the tests several rimes with different random seeds to generate the module-level test data.

ogrisel replied:
Should be fixed in 5ba0fbeea2a5c12e8f888bc20b37a0403d0ab40b.

jjerphan committed:
Do use the default rtol

ogrisel committed:
Set rtol explicitly in test_distance_metrics_dtype_consistency

ogrisel reviewed (approved):
LGTM!

ogrisel commented:
@jeremiedbb @lorentzenchr I think this PR is ready for final review.

jeremiedbb reviewed (commented):
I made a quick pass. Looks good overall. Just a couple of comments for now

jjerphan committed:
Implement the sparse-dense and the dense-sparse case for c-contiguity

Also do test for c-contiguity.

jjerphan committed:
Add validation on X and Y, accepting CSR as inputs

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -445,24 +700,20 @@ cdef class DistanceMetric{{name_suffix}}:
+            return self._pairwise_sparse_dense(X, Y)
+        if not X_is_sparse and Y_is_sparse:
+            # Swapping argument and transposing the result
+            return self._pairwise_sparse_dense(Y, X).T
```
An alternative that I've tested is to provide the order to use for the distance matrix (which would have been "F" in this case) so that the result is always "C"-ordered. Yet, the Cython methods expect the distance matrix to be "C"-ordered.

I think the best is simply to implement the last case, as done in 4f4583934c151051510ba7cead1256bb6a7dabe6.

jjerphan commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -424,6 +588,97 @@ cdef class DistanceMetric{{name_suffix}}:
+        return Darr
+
+    def _pairwise_sparse_dense(self, X, Y):
+        X_csr = X.tocsr()
```
Done in 3e3e8881d4bc1c6d11a9e473a1efd8c90bb5d1cc.

ogrisel replied:
Let's drop the `X.tocsr()` calls now.

ogrisel commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -433,36 +721,36 @@ cdef class DistanceMetric{{name_suffix}}:
-                         dtype={{DTYPE}}, order='C')
-            self.cdist(Xarr, Yarr, Darr)
-        return Darr
+            Y = check_array(Y, accept_sparse=['csr'])
```
This could be simplified to:

```python
X = check_array(X, accept_sparse=['csr'])
if Y is None:
   Y = X
else:
   Y = check_array(Y, accept_sparse=['csr'])
```

jeremiedbb replied:
Sorry I must have been misleading, I don't think DistanceMetrics should do validation on X and Y. My thinking was:
DistanceMetrics expects sparse matrices in CSR format. It's the role of the estimator using DistanceMetrics (or more directly PairwiseDistancesReduction) to make sure to pass the array in the right format. So this estimator must set ``accept_sparse=["csr"]`` in its call to ``_validate_data``.

ogrisel replied:
`DistanceMetrics.pairwise` is technically never used by estimators directly: I think all estimators that rely on `DistanceMetrics` sublclasses either go through the `PairwiseDistanceReduction` class or the `BallTree` / `KDTree` classes and all of them call into the Cython API directly, without going through the public Python API.

However `DistanceMetrics.pairwise` is part of the public Python so I think it's fine to do input validation checks here.

jeremiedbb replied:
ah yes makes sense.

Can't wait for this to be deprecated :)

jjerphan committed:
Remove left-overs

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into maint/dist-metrics-csr-support

jjerphan commented:
Follow-up on IRL discussions with @ogrisel and @jeremiedbb before I forget: currently, the dense-sparse case is handled as a special case of the sparse-sparse case (where dense vectors are seen and iterated against like sparse vectors). This allows having just two interfaces (namely `DistanceMetric.{dist_csr,rdist_csr}`) for the upcoming support of distance computations on sparse-dense, dense-sparse and sparse-sparse pair of vectors.

Still, @jeremiedbb mentioned that there might be a more performant alternatives (avoiding `jmp`, IIRC) for handling the dense-sparse case and I think it makes sense to explore it.

Should we explore alternatives in this PR? I am +0 as I think alternatives for the dense-sparse case come as additional methods rather than as modifications of the new `DistanceMetric.{dist_csr,rdist_csr}`, which could in the meantime be used for the dense-sparse case.

lorentzenchr commented:
I'm in favor of proceeding with the current implementation and improve later - if we find good improvements.

lorentzenchr commented on sklearn/metrics/_dist_metrics.pyx.tp:
```diff
@@ -375,11 +393,157 @@ cdef class DistanceMetric{{name_suffix}}:
+        const SPARSE_INDEX_TYPE_t[:] x1_indices,
+        const {{INPUT_DTYPE_t}}[:] x2_data,
+        const SPARSE_INDEX_TYPE_t[:] x2_indices,
+        const SPARSE_INDEX_TYPE_t x1_start,
+        const SPARSE_INDEX_TYPE_t x1_end,
+        const SPARSE_INDEX_TYPE_t x2_start,
+        const SPARSE_INDEX_TYPE_t x2_end,
```
It would be nice to have a description of theses somewhere in the docstrings, just once.
x_data and x_indices is the 2-d sparse array. So why to I need x_start and x_end?

jjerphan replied:
Does 731370a0fad51c50a7e95ba2a52da8c75721fd17 clarifies this?

lorentzenchr replied:
A comment was added that explains those parameters. Maybe we can do a little better, but I considered it as resolved and therefore merged.

jjerphan replied:
@ogrisel: did you mean the proposal of 731370a was insufficient, out of scope, or could be rephrased? :)

lorentzenchr reviewed (approved):
I haven't reviewed all the code in _dist_metrics.pyx.tp in detail but I find the tests good enough to trust the correctness.

jeremiedbb reviewed (approved):
LGTM

jjerphan committed:
DOC Motivate the signature for DistanceMetric.{dist_csr, rdist_csr}

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
